tree/ntuple/v7/src/RPageStorageDaos.cxx

```diff
@@ -0,0 +1,462 @@
+} // namespace
+
+
+////////////////////////////////////////////////////////////////////////////////
+
+
+ROOT::Experimental::Detail::RPageSinkDaos::RPageSinkDaos(std::string_view ntupleName, std::string_view path,
+   const RNTupleWriteOptions &options)
+   : RPageSink(ntupleName, options)
+   , fMetrics("RPageSinkRoot")
+   , fPageAllocator(std::make_unique<RPageAllocatorHeap>())
+{
+   auto args = ParseDaosPoolURI(path);
+   fPool = std::make_shared<RDaosPool>(std::get<0>(args), std::get<1>(args));
+   fContainer = std::unique_ptr<RDaosContainer>(new RDaosContainer(fPool, ntupleName, /*create =*/ true));
+}
+
+
+ROOT::Experimental::Detail::RPageSinkDaos::~RPageSinkDaos()
+{
+}
+
+
+void ROOT::Experimental::Detail::RPageSinkDaos::CreateImpl(const RNTupleModel & /* model */)
+{
+   const auto &descriptor = fDescriptorBuilder.GetDescriptor();
+   auto szHeader = descriptor.SerializeHeader(nullptr);
+   auto buffer = std::unique_ptr<unsigned char[]>(new unsigned char[szHeader]);
+   descriptor.SerializeHeader(buffer.get());
+
+   auto zipBuffer = std::unique_ptr<unsigned char[]>(new unsigned char[szHeader]);
+   auto szZipHeader = fCompressor(buffer.get(), szHeader, fOptions.GetCompression(),
+      [&zipBuffer](const void *b, size_t n, size_t o){ memcpy(zipBuffer.get() + o, b, n); } );
+   WriteNTupleHeader(zipBuffer.get(), szZipHeader, szHeader);
+}
+
+
+ROOT::Experimental::RClusterDescriptor::RLocator
+ROOT::Experimental::Detail::RPageSinkDaos::CommitPageImpl(ColumnHandle_t columnHandle, const RPage &page)
+{
+   unsigned char *buffer = reinterpret_cast<unsigned char *>(page.GetBuffer());
+   bool isAdoptedBuffer = true;
+   auto packedBytes = page.GetSize();
+   auto element = columnHandle.fColumn->GetElement();
+   const auto isMappable = element->IsMappable();
+
+   if (!isMappable) {
+      packedBytes = (page.GetNElements() * element->GetBitsOnStorage() + 7) / 8;
+      buffer = new unsigned char[packedBytes];
+      isAdoptedBuffer = false;
+      element->Pack(buffer, page.GetBuffer(), page.GetNElements());
+   }
+   auto zippedBytes = packedBytes;
+
+   if (fOptions.GetCompression() != 0) {
+      zippedBytes = fCompressor(buffer, packedBytes, fOptions.GetCompression());
+      if (!isAdoptedBuffer)
+         delete[] buffer;
+      buffer = const_cast<unsigned char *>(reinterpret_cast<const unsigned char *>(fCompressor.GetZipBuffer()));
+      isAdoptedBuffer = true;
+   }
+
+   auto offsetData = std::get<0>(fContainer->WriteObject(buffer, zippedBytes,
+                                                         kDistributionKey, kAttributeKey)).lo;
+   fClusterMinOffset = std::min(offsetData, fClusterMinOffset);
+   fClusterMaxOffset = std::max(offsetData + zippedBytes, fClusterMaxOffset);
+
+   if (!isAdoptedBuffer)
+      delete[] buffer;
+
+   RClusterDescriptor::RLocator result;
+   result.fPosition = offsetData;
+   result.fBytesOnStorage = zippedBytes;
+   return result;
+}
+
+
+ROOT::Experimental::RClusterDescriptor::RLocator
```

Comment:
Let's add a TODO comment here: the current byte range arithmetic makes little sense for the object store.  We might find out, however, that there are native ways to group clusters in DAOS.